README.md
# SurveyMonkeyTools

This is a utility tools for bulk duplicating new surveys from existing surveys.

## Installation

    $ copy .env.example .env

Replace the placeholder in `.env` with your own Access Token of SurveyMonkey App.  
Set GitHub Secrets named ACCESS_TOKEN.

## Usage

To GET all surveys list

    $ bundle exec ruby exe/survey_monkey_tools surveys

To POST survey

    $ bundle exec ruby exe/survey_monkey_tools post_survey --params='{"title":"sample title", "from_template_id":"2"}'

To PATCH survey

    $ bundle exec ruby exe/survey_monkey_tools patch_survey id --params='{"title":"new title", "folder_id:3"}'

To GET all survey folders list

    $ bundle exec ruby exe/survey_monkey_tools folders

To POST folder

    $ bundle exec ruby exe/survey_monkey_tools post_folder --params='{"title":"sample folder"}

## Features

- GET /surveys
- POST /surveys
- GET /survey_folders
- POST /survey_folders

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/Naoyuk/survey_monkey_tools. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/Naoyuk/survey_monkey_tools/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the SurveyMonkeyTools project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/Naoyuk/survey_monkey_tools/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.6.0] - 2023-09-02

- Added SurveyMonkeyTools::CLI#patch_survey
- Changed SurveyMonkeyTools::CLI#post_survey to SurveyMonkeyTools::CLI#surveys

## [0.5.0] - 2023-08-31

- Added SurveyMonkeyTools::CLI#create_survey

## [0.4.0] - 2023-08-30

- Added SurveyMonkeyTools::CLI#create_folder

## [0.3.0] - 2023-08-30

- Added SurveyMonkeyTools::CLI#folders
- Changed SurveyMonkeyTools::CLI#get to SurveyMonkeyTools::CLI#surveys

## [0.2.1] - 2023-08-30

- Update API doc

## [0.2.0] - 2023-08-30

- Added SurveyMonkeyTools::CLI#get

## [0.1.0] - 2023-08-28

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true

require "survey_monkey_tools"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end

  config.after(:suite) do
    WebMock.reset!
  end
end


spec/survey_monkey_tools_spec.rb
# frozen_string_literal: true

require "survey_monkey_tools"
require "webmock/rspec"
require "json"

RSpec.describe SurveyMonkeyTools do
  describe "#version" do
    it "output version strings" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new
      expect do
        survey_monkey_tools.version
      end.to output("survey_monkey_tools #{SurveyMonkeyTools::VERSION}\n").to_stdout
    end
  end

  describe "GET #surveys" do
    before do
      filepath = "fixtures/surveymonkey/get_surveys_response.json"
      endpoint = SurveyMonkeyTools::END_POINTS[:surveys]
      build_stub(filepath, :get, endpoint, 200)
    end

    it "returns a correct stdout" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new

      expect { survey_monkey_tools.surveys }.to output(@body).to_stdout
    end
  end

  describe "POST #post_survey" do
    before do
      filepath = "fixtures/surveymonkey/post_surveys_response.json"
      endpoint = SurveyMonkeyTools::END_POINTS[:surveys]
      build_stub(filepath, :post, endpoint, 201)
    end

    it "returns a correct stdout" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new
      filepath = "fixtures/surveymonkey/post_survey_request.json"
      params = JSON.parse(File.read(File.join(__dir__, filepath)))
      stdout = "Survey is successfully created\nStatus: 201\n#{@body}"

      expect { survey_monkey_tools.post_survey(params) }.to output(stdout).to_stdout
    end
  end

  describe "PATCH #patch_survey" do
    before do
      filepath = "fixtures/surveymonkey/patch_survey_response.json"
      @id = 1
      endpoint = "#{SurveyMonkeyTools::END_POINTS[:surveys]}/#{@id}"
      build_stub(filepath, :patch, endpoint, 200)
    end

    it "returns a correct stdout" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new
      filepath = "fixtures/surveymonkey/post_survey_request.json"
      params = JSON.parse(File.read(File.join(__dir__, filepath)))
      stdout = "Survey is successfully updated\nStatus: 200\n#{@body}"

      expect { survey_monkey_tools.patch_survey(@id, params) }.to output(stdout).to_stdout
    end
  end

  describe "GET #folders" do
    before do
      filepath = "fixtures/surveymonkey/get_folders_response.json"
      endpoint = SurveyMonkeyTools::END_POINTS[:folders]
      build_stub(filepath, :get, endpoint, 200)
    end

    it "returns a correct stdout" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new

      expect { survey_monkey_tools.folders }.to output(@body).to_stdout
    end
  end

  describe "POST #post_folder" do
    before do
      filepath = "fixtures/surveymonkey/post_folders_response.json"
      endpoint = SurveyMonkeyTools::END_POINTS[:folders]
      build_stub(filepath, :post, endpoint, 201)
    end

    it "returns a success response" do
      survey_monkey_tools = SurveyMonkeyTools::CLI.new
      stdout = "Folder is successfully created\nStatus: 201\n#{@body}"

      expect { survey_monkey_tools.post_folder("Teams Polls") }.to output(stdout).to_stdout
    end
  end

  def surveymonkey_request(method, endpoint)
    headers = {
      "Accept" => "application/json",
      "Authorization" => "Bearer #{ENV["ACCESS_TOKEN"]}"
    }
    headers["Content-Type"] = "application/json" if method == :post || method == :patch

    stub_request(method, "#{SurveyMonkeyTools::BASE_URI}/v3#{endpoint}").with(
      headers: headers
    )
  end

  def build_stub(filepath, method, endpoint, response_status)
    @body = File.read(File.join(__dir__, filepath))
    headers = {
      "Accept": "application/json",
      "Authorization": "Bearer #{ENV['ACCESS_TOKEN']}"
    }
    headers["Content-Type"] = "application/json" if method == :post || method == :patch

    surveymonkey_request(method, endpoint).to_return(
      body: @body,
      status: response_status,
      headers: headers
    )
  end
end


lib/survey_monkey_tools/cli.rb
# frozen_string_literal: true

require "survey_monkey_tools"
require "json"
require "thor"
require "uri"
require "net/http"
require "dotenv"
require "debug"

Dotenv.load

module SurveyMonkeyTools
  # class for CLI command execution
  class CLI < Thor
    desc "version", "shows survey_monkey_tools version"
    def version
      puts "survey_monkey_tools #{VERSION}"
    end

    desc "surveys", "gets surveys"
    def surveys
      puts response(END_POINTS[:surveys], :get).body
    end

    desc "post_survey", "creates a survey"
    option :params, type: :hash
    def post_survey(params)
      res = response(END_POINTS[:surveys], :post, params: params)
      message = res.code == "201" ? "Survey is successfully created" : "Error has occured"
      puts "#{message}\nStatus: #{res.code}\n#{res.body}"
    rescue StandardError => e
      puts e
    end

    desc "patch_survey", "updates a survey"
    option :id
    option :params, type: :hash
    def patch_survey(id, params)
      res = response("#{END_POINTS[:surveys]}/#{id}", :patch, params: params)
      message = res.code == "200" ? "Survey is successfully updated" : "Error has occured"
      puts "#{message}\nStatus: #{res.code}\n#{res.body}"
    rescue StandardError => e
      puts e
    end

    desc "folders", "gets folders"
    def folders
      puts response(END_POINTS[:folders], :get).body
    end

    desc "copy", "copys new surveys from surveys of last year"
    def copy
      puts "done copy"
    end

    desc "post_folder", "creates a folder"
    option :params, type: :hash
    def post_folder(params)
      res = response(END_POINTS[:folders], :post, params: params)
      message = res.code == "201" ? "Folder is successfully created" : "Error has occured"
      puts "#{message}\nStatus: #{res.code}\n#{res.body}"
    rescue StandardError => e
      puts e
    end

    private

    def access_token
      ENV["ACCESS_TOKEN"]
    end

    def http
      uri = URI.parse(BASE_URI)
      http = Net::HTTP.new(uri.host, uri.port)
      http.use_ssl = true
      http
    end

    def response(endpoint, method, params=nil)
      headers = { Accept: "application/json", Authorization: "Bearer #{access_token}" }

      case method
        when :get
          req = Net::HTTP::Get.new("/v3#{endpoint}", headers)
        when :post, :patch
          req = Net::HTTP.const_get(method.capitalize).new("/v3#{endpoint}", headers)
          req.content_type = "application/json"
          req.body = JSON.parse(params[:params]).to_json
          # req.body = params[:params].to_json
        # Uncomment and add other HTTP methods as needed
        # when :put
        # when :delete
        else
          raise ArgumentError, "Unsupported HTTP method: #{method}"
        end

      send_request(req)
    end

    def send_request(req)
      begin
        http.request(req)
      rescue StandardError => e
        puts e
      end
    end
  end
end
